### PR TITLE
add extras_require complete and test options

### LINKED ISSUE #553: clearsky.py requires PyTables but setup.py doesn't list it as install_requires
**Getting clearsky from Location causes an error**
I tried the [rdtools' excample](https://github.com/NREL/rdtools/blob/master/docs/degradation_example.ipynb), which used pvlib. 
At the first line under section "Clear Sky 0: Preliminary Calculations", executing 
clearsky = loc.get_clearsky(df.index, solar_position = sun)
causes an error:
ImportError: The Linke turbidity lookup table requires tables. You can still use clearsky.ineichen if you supply your own turbidities. 
Looking at the code found that it requires PyTables package. 
Installing PyTables by pip "pip install tables" solved the problem. 

**To Reproduce**
1. Create a Location. e.g.
   loc = pvlib.location.Location(latitude, longitude, tz = timezone)
2. Get clearsky from loc. e.g.
   clearsky = loc.get_clearsky(df.index, solar_position = sun)
3. Execute the code

**Expected behavior**
The problem is that PyTables package is required but this is not documented, nor written in setup.py. 
This should be documented and PyTables should be lisetd as install_requires in setup.py.

**Versions:**
 - ``pvlib.__version__``: 0.5.2
 - ``pandas.__version__``: 0.23.4
 - python: 3.6.4

**Additional context**
This is not necessarily a bug. But other categories didn't seem to fit thus I put this as a Bug report.

### PR DESCRIPTION
**Pull request by wholmgren**
- [x] Closes #553
 - [x] I am familiar with the [contributing guidelines](http://pvlib-python.readthedocs.io/en/latest/contributing.html).
 - [x] Fully tested. Added and/or modified tests to ensure correct behavior for all reasonable inputs. Tests (usually) must pass on the TravisCI and Appveyor testing services.
 - [x] Updates entries to `docs/sphinx/source/api.rst` for API changes.
 - [x] Adds description and name entries in the appropriate `docs/sphinx/source/whatsnew` file for all changes.
 - [x] Code quality and style is sufficient. Passes LGTM and SticklerCI checks.
 - [x] New code is fully documented. Includes sphinx/numpydoc compliant docstrings and comments in the code where necessary.
 - [x] Pull request is nearly complete and ready for detailed review.

I'm not willing to make scipy, tables, etc. requirements for the upcoming 0.6.0 release. But I think we can use the setuptools extras_require option to at least make it a little easier for a few people. So people will be able to install all of the dependencies necessary for any feature by running

*edited below here for clarity and future reference*

```
pip install pvlib[optional]  # once released to pypi
pip install -e .[optional]  # works for development too, though escape brackets for zsh
```

I *originally* chose "complete" because that's what dask uses. See discussion below for reasons to changing to optional/doc/test/all.

Options can also be combined like this:

```
pip install -e .[optional,test]
```

@mikofski you are most qualified to review this!

### CODE REVIEW

**Comment by mikofski** [comment]
#### LGTM:

1. create a virtualenv, install pvlib requirements and build a wheel

```bash
markmikofski@Suzannes-MacBook-Pro:pvlib-python (extras)$ virtualenv -p python3 venv

markmikofski@Suzannes-MacBook-Pro:pvlib-python (extras)$ . venv/bin/activate

(venv) markmikofski@Suzannes-MacBook-Pro:pvlib-python (extras)$ pip install numpy scipy siphon netcdf4 matplotlib pytest mock nose pytest-cov pandas ipython jupyter sphinx sphinx-rtd-theme numpydoc numba ephem pytest-mock tables nbsphinx pytz

(venv) markmikofski@Suzannes-MacBook-Pro:pvlib-python (extras)$ python setup.py bdist_wheel sdist

(venv) markmikofski@Suzannes-MacBook-Pro:pvlib-python (extras)$ ls dist/
pvlib-0.5.2+48.gf427a39-py3-none-any.whl
pvlib-0.5.2+48.gf427a39.tar.gz

(venv) markmikofski@Suzannes-MacBook-Pro:pvlib-python (extras)$ deactivate
```

2. create another virtualenv and install wheel with extra requirements `[complete]`

<details>

```bash
markmikofski@Suzannes-MacBook-Pro:pvlib-python (extras)$ virtualenv -p python3 venv-extras

markmikofski@Suzannes-MacBook-Pro:pvlib-python (extras)$ . venv-extras/bin/activate

(venv-extras) markmikofski@Suzannes-MacBook-Pro:pvlib-python (extras)$ pip list
Package    Version
---------- -------
pip        18.0   
setuptools 40.2.0 
wheel      0.31.1 

(venv-extras) markmikofski@Suzannes-MacBook-Pro:pvlib-python (extras)$ cd

(venv-extras) markmikofski@Suzannes-MacBook-Pro:~$ pip install --no-index --find-links ~/Projects/pvlib-python/dist/ pvlib[complete]
Looking in links: /Users/markmikofski/Projects/pvlib-python/dist/
Collecting pvlib[complete]
Collecting six (from pvlib[complete])
  Could not find a version that satisfies the requirement six (from pvlib[complete]) (from versions: )
No matching distribution found for six (from pvlib[complete])
(venv-extras) markmikofski@Suzannes-MacBook-Pro:~$ pip install --find-links ~/Projects/pvlib-python/dist/ pvlib[complete]
Looking in links: /Users/markmikofski/Projects/pvlib-python/dist/
Collecting pvlib[complete]
Collecting pytz (from pvlib[complete])
  Using cached https://files.pythonhosted.org/packages/30/4e/27c34b62430286c6d59177a0842ed90dc789ce5d1ed740887653b898779a/pytz-2018.5-py2.py3-none-any.whl
Collecting pandas>=0.14.0 (from pvlib[complete])
  Using cached https://files.pythonhosted.org/packages/78/78/50ef81a903eccc4e90e278a143c9a0530f05199f6221d2e1b21025852982/pandas-0.23.4-cp36-cp36m-macosx_10_6_intel.macosx_10_9_intel.macosx_10_9_x86_64.macosx_10_10_intel.macosx_10_10_x86_64.whl
Collecting numpy>=1.10.1 (from pvlib[complete])
  Using cached https://files.pythonhosted.org/packages/a6/0a/2defbdd9bd3a436ee0642bb8d8a66d16e523122d33384a04548dcfab23ba/numpy-1.15.1-cp36-cp36m-macosx_10_6_intel.macosx_10_9_intel.macosx_10_9_x86_64.macosx_10_10_intel.macosx_10_10_x86_64.whl
Collecting six (from pvlib[complete])
  Using cached https://files.pythonhosted.org/packages/67/4b/141a581104b1f6397bfa78ac9d43d8ad29a7ca43ea90a2d863fe3056e86a/six-1.11.0-py2.py3-none-any.whl
Collecting siphon; extra == "complete" (from pvlib[complete])
  Using cached https://files.pythonhosted.org/packages/56/f5/2be03af9ac2911d4795ac69de86dbd1b48c404a7812a4bf1b00403eafda5/siphon-0.8.0-py2.py3-none-any.whl
Collecting scipy; extra == "complete" (from pvlib[complete])
  Using cached https://files.pythonhosted.org/packages/a0/b6/70bf61c1badb5fea82d4c558e05e76c2dee5e77bb072fe465d7c7a87287d/scipy-1.1.0-cp36-cp36m-macosx_10_6_intel.macosx_10_9_intel.macosx_10_9_x86_64.macosx_10_10_intel.macosx_10_10_x86_64.whl
Collecting ephem; extra == "complete" (from pvlib[complete])
Collecting tables; extra == "complete" (from pvlib[complete])
  Using cached https://files.pythonhosted.org/packages/bc/5f/79643d50fcf927dcd4b3bd118062550fdcfbd247d63cad1dfb305c6852b5/tables-3.4.4-cp36-cp36m-macosx_10_6_intel.macosx_10_9_intel.macosx_10_9_x86_64.macosx_10_10_intel.macosx_10_10_x86_64.whl
Collecting numba; extra == "complete" (from pvlib[complete])
  Using cached https://files.pythonhosted.org/packages/fc/f6/1555a24d032c1ebe6db3e92796d5846e7d5b4337eeb53439d95fb4694fc2/numba-0.39.0-cp36-cp36m-macosx_10_9_x86_64.whl
Collecting python-dateutil>=2.5.0 (from pandas>=0.14.0->pvlib[complete])
  Using cached https://files.pythonhosted.org/packages/cf/f5/af2b09c957ace60dcfac112b669c45c8c97e32f94aa8b56da4c6d1682825/python_dateutil-2.7.3-py2.py3-none-any.whl
Collecting requests>=1.2 (from siphon; extra == "complete"->pvlib[complete])
  Using cached https://files.pythonhosted.org/packages/65/47/7e02164a2a3db50ed6d8a6ab1d6d60b69c4c3fdf57a284257925dfc12bda/requests-2.19.1-py2.py3-none-any.whl
Collecting beautifulsoup4>=4.6 (from siphon; extra == "complete"->pvlib[complete])
  Using cached https://files.pythonhosted.org/packages/21/0a/47fdf541c97fd9b6a610cb5fd518175308a7cc60569962e776ac52420387/beautifulsoup4-4.6.3-py3-none-any.whl
Collecting protobuf>=3.0.0a3 (from siphon; extra == "complete"->pvlib[complete])
  Using cached https://files.pythonhosted.org/packages/c7/27/133f225035b9539f2dcfebcdf9a69ff0152f56e0120160ec5c972ea7deb9/protobuf-3.6.1-cp36-cp36m-macosx_10_6_intel.macosx_10_9_intel.macosx_10_9_x86_64.macosx_10_10_intel.macosx_10_10_x86_64.whl
Collecting numexpr>=2.5.2 (from tables; extra == "complete"->pvlib[complete])
  Using cached https://files.pythonhosted.org/packages/53/e4/4ea4e519ddbf8b2c26edccfba6a6011419208044c07f8e4f01b3fb3028ba/numexpr-2.6.8-cp36-cp36m-macosx_10_6_intel.macosx_10_9_intel.macosx_10_9_x86_64.macosx_10_10_intel.macosx_10_10_x86_64.whl
Collecting llvmlite>=0.24.0dev0 (from numba; extra == "complete"->pvlib[complete])
  Using cached https://files.pythonhosted.org/packages/e7/ee/9e827a16b0f168b4e2da3efc8e3798fc3e4139f3196dafdf127b052c683d/llvmlite-0.24.0-cp36-cp36m-macosx_10_9_x86_64.whl
Collecting idna<2.8,>=2.5 (from requests>=1.2->siphon; extra == "complete"->pvlib[complete])
  Using cached https://files.pythonhosted.org/packages/4b/2a/0276479a4b3caeb8a8c1af2f8e4355746a97fab05a372e4a2c6a6b876165/idna-2.7-py2.py3-none-any.whl
Collecting urllib3<1.24,>=1.21.1 (from requests>=1.2->siphon; extra == "complete"->pvlib[complete])
  Using cached https://files.pythonhosted.org/packages/bd/c9/6fdd990019071a4a32a5e7cb78a1d92c53851ef4f56f62a3486e6a7d8ffb/urllib3-1.23-py2.py3-none-any.whl
Collecting certifi>=2017.4.17 (from requests>=1.2->siphon; extra == "complete"->pvlib[complete])
  Using cached https://files.pythonhosted.org/packages/df/f7/04fee6ac349e915b82171f8e23cee63644d83663b34c539f7a09aed18f9e/certifi-2018.8.24-py2.py3-none-any.whl
Collecting chardet<3.1.0,>=3.0.2 (from requests>=1.2->siphon; extra == "complete"->pvlib[complete])
  Using cached https://files.pythonhosted.org/packages/bc/a9/01ffebfb562e4274b6487b4bb1ddec7ca55ec7510b22e4c51f14098443b8/chardet-3.0.4-py2.py3-none-any.whl
Requirement already satisfied: setuptools in ./Projects/pvlib-python/venv-extras/lib/python3.6/site-packages (from protobuf>=3.0.0a3->siphon; extra == "complete"->pvlib[complete]) (40.2.0)
Installing collected packages: pytz, six, python-dateutil, numpy, pandas, idna, urllib3, certifi, chardet, requests, beautifulsoup4, protobuf, siphon, scipy, ephem, numexpr, tables, llvmlite, numba, pvlib
Successfully installed beautifulsoup4-4.6.3 certifi-2018.8.24 chardet-3.0.4 ephem-3.7.6.0 idna-2.7 llvmlite-0.24.0 numba-0.39.0 numexpr-2.6.8 numpy-1.15.1 pandas-0.23.4 protobuf-3.6.1 pvlib-0.5.2+48.gf427a39 python-dateutil-2.7.3 pytz-2018.5 requests-2.19.1 scipy-1.1.0 siphon-0.8.0 six-1.11.0 tables-3.4.4 urllib3-1.23
(venv-extras) markmikofski@Suzannes-MacBook-Pro:~$ pip list
Package         Version          
--------------- -----------------
beautifulsoup4  4.6.3            
certifi         2018.8.24        
chardet         3.0.4            
ephem           3.7.6.0          
idna            2.7              
llvmlite        0.24.0           
numba           0.39.0           
numexpr         2.6.8            
numpy           1.15.1           
pandas          0.23.4           
pip             18.0             
protobuf        3.6.1            
pvlib           0.5.2+48.gf427a39
python-dateutil 2.7.3            
pytz            2018.5           
requests        2.19.1           
scipy           1.1.0            
setuptools      40.2.0           
siphon          0.8.0            
six             1.11.0           
tables          3.4.4            
urllib3         1.23             
wheel           0.31.1           
```

</details><br/>

**Comment by mikofski** [comment]
odd I don't see `netcdf4` in the virtualenv with the "complete" install, but it is a requirement of the building the [docs](https://github.com/pvlib/pvlib-python/blob/fe5d90d3aa068d549f17cb326dd24faab177c8b6/docs/environment.yml) and the [`forecast.py`](https://github.com/pvlib/pvlib-python/blob/f16e53e2763c27fa614eb5f285edc3abed956826/pvlib/forecast.py) module

probably best to add it

File: setup.py [hunk 1]
```diff
@@ -42,7 +42,11 @@
                     'pytz',
                     'six',
                     ]
-TESTS_REQUIRE = ['pytest', 'nose']
+TESTS_REQUIRE = ['pytest', 'pytest-cov', 'pytest-mock', 'nose']
+EXTRAS_REQUIRE = {
+    'complete': ['scipy', 'tables', 'numba', 'siphon', 'ephem'],
```
Comment (mikofski) [inline setup.py]: please add `netcdf4`

Comment (wholmgren) [reply setup.py]: thanks for catching this. working on adding a ``doc`` section, too.

File: setup.py [hunk 2]
```diff
@@ -44,7 +44,9 @@
                     ]
 TESTS_REQUIRE = ['pytest', 'pytest-cov', 'pytest-mock', 'nose']
 EXTRAS_REQUIRE = {
-    'complete': ['scipy', 'tables', 'numba', 'siphon', 'ephem'],
+    'complete': ['scipy', 'tables', 'numba', 'siphon', 'netcdf4', 'ephem'],
+    'doc': ['sphinx', 'ipython', 'sphinx_rtd_theme', 'numpydoc',
+            'matplotlib', 'jupyter'],
```
Comment (mikofski) [inline setup.py]: I believe `nbsphinx` is also necessary for docs

**Comment by mikofski** [comment]
for reference, here's my [`requirements.txt`](../pull/533#discussion_r209776825)

**Comment by wholmgren** [comment]
Here's a link to the build: https://wholmgren-pvlib-python-new.readthedocs.io/en/extras/index.html

but none of the changes in this PR should effect the documentation build on readthedocs.

I'm wondering if we'd be better off with something like

```
pvlib[optional]  # scipy, tables, etc.
pvlib[doc]  # sphinx, etc
pvlib[test]  # pytest etc
pvlib[complete] or pvlib[all]  # all of the above, same as pvlib[optional,doc,test]
```

**Comment by mikofski** [comment]
1. Yep, sorry fir cross posting
2. Great idea. Sounds good to me.

**Comment by wholmgren** [comment]
Ok @mikofski, now the changes in this PR do effect the doc build! Between wheels and readthedocs.org including numpy/scipy/matplotlib in its root environment, I think we're now better off using pip installs on RTD than conda envs. This PR consolidates the build requirements in one place (``setup.py``) and the builds seem to be faster (4-8 minutes instead of 12-15+).

Unfortunately, reconfiguring RTD has not fixed the ipython directive issue.

**Comment by mikofski** [comment]
At some point does it make sense asking on [Google groups](https://groups.google.com/forum/m/#!forum/read-the-docs) make sense? Or searching [SO](https://stackoverflow.com/questions/tagged/read-the-docs)?

**Comment by wholmgren** [comment]
Well I had issues with it locally so not sure what to ask specific to rtd.

**Comment by wholmgren** [comment]
@mikofski are you happy with the latest changes here? You are the active contributor that is most likely to take advantage of this on a regular basis, so want to make sure you're happy with it. I'm ready to merge if you still approve. 

The [latest](https://pvlib-python.readthedocs.io/en/latest/timetimezones.html) doc build (still on conda) now shows the same ipython behavior as in the build of my fork's branch. I think we should go ahead with changing the pvlib RTD configuration to use wheels.

**Comment by mikofski** [comment]
LGTM, if I have a chance, I'll try to figure out what's going on with ipython  directive, maybe there's an alterate way to do the doctest, but for now I think this is fine. Thanks again! OK with me to merge ...

**Pull request merged**